contracts/swap/src/queries.rs

```diff
@@ -162,140 +234,166 @@ fn estimate_execution_buy(
-    let expected_quantity = expected_exchange_quantity - expected_fee;
-    deps.api.debug(&format!("Sell exchange: {expected_exchange_quantity}, fee: {expected_fee}, total: {expected_quantity}"));
-    let worst_price = worst_price(&top_orders);
+    let top_orders =
+        get_minimum_liquidity_levels(deps, &orders.buys_price_level, available_funds, |l| {
+            l.q * l.p
+        })?;
+
+    let average_price = get_average_price_from_orders(&top_orders);
+    let expected_input_quantity = amount_coin.amount / average_price;
+    let worst_price = get_worst_price_from_orders(&top_orders);
 
     Ok(StepExecutionEstimate {
         worst_price,
-        result_quantity: round_to_min_tick(expected_quantity, market.min_price_tick_size),
+        result_quantity: round_to_min_tick(expected_input_quantity, market.min_price_tick_size),
         result_denom: counter_denom(market, &amount_coin.denom)?.to_string(),
         is_buy_order: false,
         fee_estimate: Some(FPCoin {
             denom: market.quote_denom.clone(),
-            amount: expected_fee,
+            amount: fee_estimate,
         }),
     })
 }
 
-pub fn find_minimum_orders(
+fn estimate_execution_sell(
     deps: &Deps<InjectiveQueryWrapper>,
+    querier: &InjectiveQuerier,
+    market: &SpotMarket,
+    swap_estimation_amount: SwapEstimationAmount,
+    fee_percent: FPDecimal,
+) -> StdResult<StepExecutionEstimate> {
+    let amount_coin = match swap_estimation_amount.to_owned() {
+        SwapEstimationAmount::InputQuantity(fp) => fp,
+        SwapEstimationAmount::ReceiveQuantity(fp) => fp,
+    };
+
+    let is_estimating_from_target = matches!(
+        swap_estimation_amount,
+        SwapEstimationAmount::ReceiveQuantity(_)
+    );
+
+    if is_estimating_from_target {
+        estimate_execution_sell_from_target(deps, querier, market, amount_coin, fee_percent)
+    } else {
+        estimate_execution_sell_from_source(deps, querier, market, amount_coin, fee_percent)
+    }
+}
+
+pub fn get_minimum_liquidity_levels(
+    _deps: &Deps<InjectiveQueryWrapper>,
     levels: &Vec<PriceLevel>,
     total: FPDecimal,
     calc: fn(&PriceLevel) -> FPDecimal,
 ) -> StdResult<Vec<PriceLevel>> {
-    deps.api
-        .debug(&format!("find_minimum_orders, total: {total}"));
-    deps.api.debug(&format!("levels: {levels:?}",));
     let mut sum = FPDecimal::zero();
     let mut orders: Vec<PriceLevel> = Vec::new();
+
     for level in levels {
         let value = calc(level);
-        deps.api.debug(&format!(
-            "Adding level {}x{} value: {}, sum so far: {sum}",
-            level.p.scaled(12),
-            level.q.scaled(-18),
-            value.scaled(-18),
-        ));
+
         let order_to_add = if sum + value > total {
             let excess = value + sum - total;
-            deps.api.debug(&format!(
-                "Value: {value}, excess value: {excess}, sum so far: {}",
-                sum.scaled(-18)
-            ));
```

Comment:
in general I think it's a bad idea to remove debug logs. In 10 minutes we will be adding them back to debug something...

Reply:
It makes the code completely so much more unreadable. I think we should first and foremost have clean clode -> less bugs to debug to begin with.

That said, there may be ways to make everyone happy e.g. by strategically placing debug logs only in one spot and keeping it to an absolute minimum? 🤷

Reply:
We can add them back in if required, but following my suggestions above.